### PR TITLE
Move headers under detail subfolder and use modern cmake to fix single typesupport builds

### PR DESCRIPTION
**Pull request by ivanpauno**
Depends on https://github.com/ros2/rosidl/pull/483.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Please post CI builds before / after.

**Comment by ivanpauno** [comment]
This will still fail, but running before and after jobs to show the difference:

* Before [Build Status](http://ci.ros2.org/job/ci_linux/10663/)
* After [Build Status](http://ci.ros2.org/job/ci_linux/10662/)

**Comment by ivanpauno** [comment]
Depends on https://github.com/ros2/rosidl/pull/483

**Comment by ivanpauno** [comment]
* After [Build Status](http://ci.ros2.org/job/ci_linux/10667/)

If I delete the `QUIET` here https://github.com/ros2/rosidl/blob/f3f639df4cef29fda1dc18074fe0b1e2c3b108e6/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake#L145, the error is:

```
CMake Warning at /home/ivanpauno/ros2_ws/other_ws/investigating_build_connext_failure/install/rosidl_cmake/share/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake:145 (find_package):
  Found package configuration file:

    /home/ivanpauno/ros2_ws/other_ws/investigating_build_connext_failure/install/action_msgs/share/action_msgs/cmake/action_msgsConfig.cmake

  but it set action_msgs_FOUND to FALSE so package "action_msgs" is
  considered to be NOT FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing:
  action_msgs::action_msgs__rosidl_typesupport_fastrtps_cpp
```

The problem is that though the target is correctly exported, `action_msgs__rosidl_typesupport_fastrtps_c` depends on it and it's included before.

**Comment by ivanpauno** [comment]
* Before [Build Status](http://ci.ros2.org/job/ci_linux/10780/)
* After [Build Status](http://ci.ros2.org/job/ci_linux/10781/)

**Comment by ivanpauno** [comment]
Multiple typesupports build:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10782/)

**Comment by ivanpauno** [comment]
@dirk-thomas PTAL

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake [hunk 1]
```diff
@@ -26,8 +32,8 @@ foreach(_abs_idl_file ${rosidl_generate_interfaces_ABS_IDL_FILES})
   get_filename_component(_idl_name "${_abs_idl_file}" NAME_WE)
   string_camel_case_to_lower_case_underscore("${_idl_name}" _header_name)
   list(APPEND _generated_files
-    "${_output_path}/${_parent_folder}/${_header_name}__rosidl_typesupport_fastrtps_c.h"
-    "${_output_path}/${_parent_folder}/${_header_name}__type_support_c.cpp")
+    "${_output_path}/${_parent_folder}/detail/${_header_name}__rosidl_typesupport_fastrtps_c.h"
+    "${_output_path}/${_parent_folder}/detail/${_header_name}__type_support_c.cpp")
```
Comment (dirk-thomas) [inline rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:36]: This change seems kind of unrelated to the topic of the PR.

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake [hunk 2]
```diff
@@ -181,13 +182,14 @@ if(NOT rosidl_generate_interfaces_SKIP_INSTALL)
 
   install(
     TARGETS ${rosidl_generate_interfaces_TARGET}${_target_suffix}
+    EXPORT ${rosidl_generate_interfaces_TARGET}${_target_suffix}
     ARCHIVE DESTINATION lib
     LIBRARY DESTINATION lib
     RUNTIME DESTINATION bin
   )
-
-  rosidl_export_typesupport_libraries(${_target_suffix}
```
Comment (dirk-thomas) [inline rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:189]: I am not sure this should be removed. In other PRs related to modern CMake we have kept the classic CMake variables to avoid breaking downstream packages.

File: rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c-extras.cmake.in [hunk 3]
```diff
@@ -11,6 +11,10 @@ if(NOT FastRTPS_FOUND)
     "Could not find eProsima Fast-RTPS: skipping rosidl_typesupport_fastrtps_c"
   )
 else()
+  # Make sure extension points are registered in order
+  find_package(rosidl_generator_c QUIET)
+  find_package(rosidl_typesupport_fastrtps_cpp QUIET REQUIRED)
```
Comment (dirk-thomas) [inline rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c-extras.cmake.in:16]: Can you elaborate why (only) one is required?

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake (hunk 1 above)
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:36]: From the [Before](http://ci.ros2.org/job/ci_linux/10780/) build above:

```
13:47:15 compilation terminated.
13:47:15 [K/home/jenkins-agent/workspace/ci_linux/ws/build/rosidl_generator_py/rosidl_typesupport_c/rosidl_generator_py/msg/bounded_sequences__type_support.cpp:10:[Kfatal erro[Krosidl_generator_py/msg/detail/bounded_sequences__rosidl_typesupport_fastrtps_c.h: No such file or directory
13:47:15    10 | #include [K"rosidl_generator_py/msg/detail/bounded_sequences__rosidl_typesupport_fastrtps_c[K
13:47:15       |          [K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
13:47:15 compilation terminated.
```

Moving the headers under a `detail` folder is not unrelated (if that is what you were pointing out).

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake [hunk 4]
```diff
@@ -123,15 +129,10 @@ set_target_properties(${rosidl_generate_interfaces_TARGET}${_target_suffix}
...
   PUBLIC
-  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c
-  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_cpp
-  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_c
-  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_cpp
-)
-ament_target_dependencies(${rosidl_generate_interfaces_TARGET}${_target_suffix}
-  "fastrtps"
-  "rosidl_typesupport_fastrtps_cpp"
-  "rosidl_typesupport_fastrtps_c"
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:134]: that was deleted by accident, I will revert that change.

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake (hunk 2 above)
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:189]: I imitated what it was done in `rosidl_typesupport_introspection_c/cpp`, where the old library export was deleted (see [1](https://github.com/ros2/rosidl/blob/c6325824cc7d52c5acb3a5662523b041afc38e8b/rosidl_typesupport_introspection_c/cmake/rosidl_typesupport_introspection_c_generate_interfaces.cmake#L150-L152), [2](https://github.com/ros2/rosidl/pull/478)).

This function didn't add the library to the `<prj_name>_LIBRARIES` variable, but instead to `<prj_name>_LIBRARIES${_lib_suffix}` that was only used by the same extension point in downstream interfaces. Thus, I don't see much value of keeping the old `rosidl_export_typesupport_libraries` call (though it can be re-added).

File: rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c-extras.cmake.in (hunk 3 above)
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c-extras.cmake.in:16]: `rosidl_generator_c` find-pkg is not required, for the same reason that we didn't want to add a runtime dependency on `rosidl_generator_c/cpp` in `rosidl_typesupport_introspection_c/cpp` (see [1](https://github.com/ros2/rosidl/pull/485#issuecomment-627534681), [2](https://github.com/ros2/rosidl/pull/485#issuecomment-627676588), [3](https://github.com/ros2/rosidl/pull/485#issuecomment-627999804)).

`rosidl_typesupport_fastrtps_cpp` find-pkg is required, because this package already has a runtime dependency on it https://github.com/ros2/rosidl_typesupport_fastrtps/blob/46b1d3dab62edd123886af3f34b96ed621cf62e4/rosidl_typesupport_fastrtps_c/package.xml#L25

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:134]: Now I remember it wasn't an accident, the original file had that repeated:

https://github.com/ros2/rosidl_typesupport_fastrtps/blob/46b1d3dab62edd123886af3f34b96ed621cf62e4/rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake#L102-L108

https://github.com/ros2/rosidl_typesupport_fastrtps/blob/46b1d3dab62edd123886af3f34b96ed621cf62e4/rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake#L131-L135

and I cleaned up one.

File: rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py [hunk 5]
```diff
@@ -18,7 +18,7 @@
 def generate_typesupport_fastrtps_c(generator_arguments_file):
     mapping = {
         'idl__rosidl_typesupport_fastrtps_c.h.em':
-        '%s__rosidl_typesupport_fastrtps_c.h',
-        'idl__type_support_c.cpp.em': '%s__type_support_c.cpp',
+        'detail/%s__rosidl_typesupport_fastrtps_c.h',
```
Comment (hidmic) [inline rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: @ivanpauno why introducing a `detail` subdirectory/namespace in this patch? Was it inconsistent?

Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: https://github.com/ros2/rosidl_typesupport_fastrtps/pull/40#discussion_r426596456

There were similar changes in `rosidl_typesupport_c/cpp` after the API review, and similar changes are needed here when a static typesupport is used.

Comment (dirk-thomas) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: > and similar changes are needed here when a static typesupport is used

The header is specific to this typesupport. That is why I think this change is unrelated and not necessary to fix the build (as per title). Have you tried to separate these changes and the build which this PR is fixing started to fail again?

Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: https://github.com/ros2/rosidl_typesupport/blob/dea9825fb7af51a74007ff5999aec544ae4c921e/rosidl_typesupport_c/resource/msg__type_support.cpp.em#L112

<details>
<summary>This is the result of a file generated by `rosidl_typesupport_c`</summary>

```cpp
// generated from rosidl_typesupport_c/resource/idl__type_support.cpp.em
// with input from test_msgs:msg/Arrays.idl
// generated code does not contain a copyright notice

#include "cstddef"
#include "rosidl_runtime_c/message_type_support_struct.h"
#include "test_msgs/msg/rosidl_typesupport_c__visibility_control.h"
#include "test_msgs/msg/detail/arrays__struct.h"
#include "rosidl_typesupport_c/visibility_control.h"
#include "test_msgs/msg/detail/arrays__rosidl_typesupport_fastrtps_c.h"

#ifdef __cplusplus
extern "C"
{
#endif

ROSIDL_TYPESUPPORT_C_EXPORT_test_msgs
const rosidl_message_type_support_t *
ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(rosidl_typesupport_c, test_msgs, msg, Arrays)() {
  return ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(rosidl_typesupport_fastrtps_c, test_msgs, msg, Arrays)();
}

#ifdef __cplusplus
}
#endif
```

</details>

Notice this line:

```cpp
#include "test_msgs/msg/detail/arrays__rosidl_typesupport_fastrtps_c.h"
```

---

I can split this in two commits (or two PRs):

* One moving headers under `detail` directory (besides fixing the build, I think this was desired after last API review).
* One doing the necessary modern cmake changes to fix the build.

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake (hunk 1 above)
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:36]: I think the discussion can be continued here https://github.com/ros2/rosidl_typesupport_fastrtps/pull/40#discussion_r427597651, so I'm resolving this other one.

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:134]: @dirk-thomas Is this change ok or should it be part of a different PR?
I included it here as it is a minimal cleanup.

File: rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py (hunk 5 above)
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: A similar change was recently done in the introspection typesupports to fix the "single typesupport" builds: https://github.com/ros2/rosidl/pull/470.

Comment (dirk-thomas) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: I leave it up to you. Imo sorting is preferred. If not, you might want to update the title to mention the detail namespace.

Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c/__init__.py:21]: I split the PR in two commits (https://github.com/ros2/rosidl_typesupport_fastrtps/pull/40/commits/5f6f49fd7697462b79bd1aad84a315b1bae18073, https://github.com/ros2/rosidl_typesupport_fastrtps/pull/40/commits/61f50aeede3eb6c8a4b43480faee2d5c22d89b5c) which I'm planing to rebase merge.

I've also updated the PR title for clarity.

File: rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake:134]: I've left this change as part of the second commit https://github.com/ros2/rosidl_typesupport_fastrtps/pull/40/commits/61f50aeede3eb6c8a4b43480faee2d5c22d89b5c.

**Comment by ivanpauno** [comment]
Job config:

build: `--packages-up-to test_msgs`
test: `--packages-select rosidl_typesupport_fastrtps_c rosidl_typesupport_fastrtps_cpp`

Single typesupport builds:

* Before [Build Status](http://ci.ros2.org/job/ci_linux/10780/)
* After [Build Status](http://ci.ros2.org/job/ci_linux/10844/)

Multiple typesupport builds:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10845/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6216/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8834/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10736/)
